sycl/include/syclcompat/traits.hpp

```diff
@@ -41,4 +42,57 @@ template <typename T> struct arith {
 };
 template <typename T> using arith_t = typename arith<T>::type;
 
+// Traits to check device function signature matches args (with or without local
+// mem)
+// TODO: naming
+template <auto F, typename... Args>
+constexpr inline bool invocable =
+    std::is_invocable<decltype(F), Args...>::value;
+
+template <auto F, typename... Args>
+constexpr inline bool lmem_invocable =
+    std::is_invocable<decltype(F), Args..., char *>::value;
+
+template <auto F, typename... Args>
+constexpr inline bool args_compatible =
+    invocable<F, Args...> || lmem_invocable<F, Args...>;
```

Comment:
Yes! It's an edge case but still I'm not happy about it. If the `launch_strategy` object had compile-time info about whether or not local memory was requested, we could avoid the ambiguity. It could have a `bool localMem` template parameter, and we could deduce it based on choice of constructor, but then if a user explicitly requests `0` bytes local memory, do we expect `char *` or no local mem arg?

Reply:
I would guess that is undefined behaviour. Anyway I am okay with going forward as is, though in the long term it may cause some bugs.

Reply:
I'm working on something which I hope will resolve this, with other things. I'm going to leave this topic unresolved to ensure I come back to it :+1:

Reply:
In the latest version, `launch_policy` has a `bool LocalMem` template parameter. This is deduced by the new variadic ctor - if there is a `compat_exp::local_mem_size` struct in the variadic ctor args, `localMem = true`. I have redefined `args_compatible` to be stricter:
```cpp
template <typename LaunchPolicy, auto F, typename... Args>
constexpr inline bool args_compatible =
    std::conditional_t<LaunchPolicy::HasLocalMem, device_fn_lmem_invocable<F, Args...>,
                     device_fn_invocable<F, Args...>>::value;
```

So now we can't slip an erroneous `char *` arg on the end

Thanks for the suggestion!